hazelcast/include/hazelcast/client/serialization/pimpl/compact.i.h

```diff
@@ -289,6 +316,55 @@ compact_reader::read_nullable_primitive(
     throw unexpected_field_kind(field_descriptor.field_kind, field_name);
 }
 
+template<typename T>
+boost::optional<std::vector<boost::optional<T>>>
+compact_reader::read_array_of_nullable(
+  const std::string& field_name,
+  enum pimpl::field_kind field_kind,
+  enum pimpl::field_kind nullable_field_kind)
+{
+    auto& field_descriptor = get_field_descriptor(field_name);
+    if (field_descriptor.field_kind == field_kind) {
+        return read_primitive_array_as_nullable_array<T>(field_descriptor);
+    } else if (field_descriptor.field_kind == nullable_field_kind) {
+        return read_array_of_variable_size<T>(field_descriptor);
+    }
+    throw unexpected_field_kind(field_descriptor.field_kind, field_name);
+}
+
+template<typename T>
+boost::optional<std::vector<boost::optional<T>>>
+compact_reader::read_primitive_array_as_nullable_array(
+  const pimpl::field_descriptor& field_descriptor)
+{
+    int32_t current_pos = object_data_input.position();
+    util::finally set_position_back(
+      [this, current_pos]() { this->object_data_input.position(current_pos); });
+
+    int32_t pos = read_var_size_position(field_descriptor);
+    if (pos == pimpl::offset_reader::NULL_OFFSET) {
+        return boost::none;
+    }
+    object_data_input.position(pos);
+    int32_t item_count = object_data_input.read<int32_t>();
+    auto values = boost::make_optional<std::vector<boost::optional<T>>>(
+      std::vector<boost::optional<T>>(item_count));
+
+    for (int i = 0; i < item_count; ++i) {
+        values.value()[i] = boost::make_optional(object_data_input.read<T>());
+    }
+    return values;
+}
+
+template<>
+boost::optional<std::vector<boost::optional<bool>>> inline compact_reader::
+  read_primitive_array_as_nullable_array(
+    const pimpl::field_descriptor& field_descriptor)
+{
+    return read_variable_size<std::vector<boost::optional<bool>>>(
+      field_descriptor);
+}
+
```

Comment:
🤔 Do we need this? There should not be a need of read_primitive_array_as_nullable_array for nullable boolean

Reply:
Yes, I think we need it. 

getArrayOfNullableBoolean calls getVariableSize in java if field_kind is `ARRAY_OF_BOOLEAN`. 

To have the same behavior. `read_array_of_nullable` should call `read_array_of_nullable` when T is `std::vector<boost::optional<bool>`.